lib/core/mergeConfig.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+var utils = require('../utils');
+
+/**
+ * Config-specific merge-function which creates a new config-object
+ * based on given defaults and instance config.
+ *
+ * @param {Object} defaults Defaults
+ * @param {Object} instanceConfig Instance-specific config
+ * @returns {Object} New object resulting from merging instanceConfig to defaults
+ */
+module.exports = function mergeConfig(defaults, instanceConfig) {
+  // eslint-disable-next-line no-param-reassign
+  instanceConfig = instanceConfig || {};
+  var config = {};
+
+  utils.forEach(['url', 'method', 'params', 'data'], function valueFromInstanceConfig(prop) {
+    config[prop] = instanceConfig[prop];
+  });
+
+  utils.forEach(['headers', 'auth', 'proxy'], function mergeInstanceConfigWithDefaults(prop) {
+    if (typeof instanceConfig[prop] !== 'undefined') {
+      if (typeof instanceConfig[prop] === 'object') {
+        config[prop] = utils.deepMerge(defaults[prop], instanceConfig[prop]);
+      } else {
+        config[prop] = instanceConfig[prop];
+      }
+    } else if (typeof defaults[prop] !== 'undefined') {
+      config[prop] = utils.deepMerge(defaults[prop]);
+    }
+  });
+
+  utils.forEach([
```

Comment:
Instead of listing all these properties here, we could apply this function to all properties that don't belong to the first two groups. Or maybe, it's actually a great idea to specify all properties explicitly so that every time a new property is added, we will have to think whether its values should be merged or not. What do you think?

Reply:
I would keep this explicit listing, that was also partly the idea behind this config-specific merge-function. The idea is that it's no longer a mystery where config-options originate from, and as you said, when adding new properties it forces to consider how they should be merged (or not).